File: impl/ts/src/skill.ts
import { CborMap, CborValue, decodeCanonical, encodeCanonical } from "./cbor.js";
import { sha256, signEd25519Hash, verifyEd25519Hash } from "./crypto.js";

const SKILL_SIG_KEY = 16;
const SANDBOX_MIN = 1;
const SANDBOX_MAX = 5;

export interface SkillArtifact {
  kind: number;
  digest: Uint8Array;
  size: number;
  uris: string[];
}

export interface SkillManifestPayload {
  skillId: string;
  author: string;
  name: string;
  version: string;
  summary: string;
  license: string;
  capabilities: string[];
  permissions: string[];
  sandboxClass: number;
  endpoints?: string[];
  artifacts?: SkillArtifact[];
  requirements?: string[];
  pricing?: CborValue;
  attestations?: CborValue;
  metadata?: CborValue;
  ts: number;
}

export interface SkillManifest {
  payload: SkillManifestPayload;
  signature: Uint8Array;
}

export interface SkillPublishPayload {
  manifest: Uint8Array;
  ts: number;
}

export interface SkillUpdatePayload {
  skillId: string;
  prevManifestHash: Uint8Array;
  manifest: Uint8Array;
  ts: number;
}

export interface SkillRevokePayload {
  skillId: string;
  manifestHash: Uint8Array;
  reason: string;
  ts: number;
}

export function parseSkillManifestPayload(value: CborValue): SkillManifestPayload {
  const entries = expectMap(value);
  const payload: SkillManifestPayload = {
    skillId: expectText(getRequired(entries, 0)),
    author: expectText(getRequired(entries, 1)),
    name: expectText(getRequired(entries, 2)),
    version: expectText(getRequired(entries, 3)),
    summary: expectText(getRequired(entries, 4)),
    license: expectText(getRequired(entries, 5)),
    capabilities: expectTextArray(getRequired(entries, 6)),
    permissions: expectTextArray(getRequired(entries, 7)),
    sandboxClass: expectU16(getRequired(entries, 8)),
    endpoints: getOptional(entries, 9) ? expectTextArray(getRequired(entries, 9)) : undefined,
    artifacts: getOptional(entries, 10) ? parseArtifacts(getRequired(entries, 10)) : undefined,
    requirements: getOptional(entries, 11) ? expectTextArray(getRequired(entries, 11)) : undefined,
    pricing: getOptional(entries, 12) ?? undefined,
    attestations: getOptional(entries, 13) ?? undefined,
    metadata: getOptional(entries, 14) ?? undefined,
    ts: expectU64(getRequired(entries, 15)),
  };
  validatePayload(payload);
  return payload;
}

export function parseSkillManifest(value: CborValue): SkillManifest {
  const [payload, signature] = splitSignedMap(value, SKILL_SIG_KEY);
  return { payload: parseSkillManifestPayload(payload), signature };
}

export function decodeSkillManifest(data: Uint8Array): SkillManifest {
  return parseSkillManifest(decodeCanonical(data));
}

export function buildSkillManifest(payload: SkillManifestPayload, secretKey: Uint8Array): Uint8Array {
  validatePayload(payload);
  const payloadMap = payloadToCbor(payload);
  const payloadBytes = encodeCanonical(payloadMap);
  const digest = sha256(payloadBytes);
  const signature = signEd25519Hash(secretKey, digest);
  const full = withSignature(payloadMap, SKILL_SIG_KEY, signature);
  return encodeCanonical(full);
}

export function verifySkillManifest(data: Uint8Array, publicKey: Uint8Array): SkillManifestPayload {
  const value = decodeCanonical(data);
  const [payload, signature] = splitSignedMap(value, SKILL_SIG_KEY);
  const payloadBytes = encodeCanonical(payload);
  const digest = sha256(payloadBytes);
  verifyEd25519Hash(publicKey, digest, signature);
  return parseSkillManifestPayload(payload);
}

export function parseSkillPublishPayload(value: CborValue): SkillPublishPayload {
  const entries = expectMap(value);
  const manifest = expectBytes(getRequired(entries, 0));
  const ts = expectU64(getRequired(entries, 1));
  if (ts === 0) throw new Error("timestamp required");
  decodeSkillManifest(manifest);
  return { manifest, ts };
}

export function parseSkillUpdatePayload(value: CborValue): SkillUpdatePayload {
  const entries = expectMap(value);
  const skillId = expectText(getRequired(entries, 0));
  const prevManifestHash = expectBytesLen(getRequired(entries, 1), 32);
  const manifest = expectBytes(getRequired(entries, 2));
  const ts = expectU64(getRequired(entries, 3));
  if (ts === 0) throw new Error("timestamp required");
  decodeSkillManifest(manifest);
  return { skillId, prevManifestHash, manifest, ts };
}

export function parseSkillRevokePayload(value: CborValue): SkillRevokePayload {
  const entries = expectMap(value);
  const skillId = expectText(getRequired(entries, 0));
  const manifestHash = expectBytesLen(getRequired(entries, 1), 32);
  const reason = expectText(getRequired(entries, 2));
  const ts = expectU64(getRequired(entries, 3));
  if (ts === 0) throw new Error("timestamp required");
  if (!reason.trim()) throw new Error("reason required");
  return { skillId, manifestHash, reason, ts };
}

export function skillPublishPayloadToCbor(payload: SkillPublishPayload): CborValue {
  if (payload.ts === 0) throw new Error("timestamp required");
  decodeSkillManifest(payload.manifest);
  return { entries: [[0, payload.manifest], [1, payload.ts]] } satisfies CborMap;
}

export function skillUpdatePayloadToCbor(payload: SkillUpdatePayload): CborValue {
  if (payload.ts === 0) throw new Error("timestamp required");
  ensureNonempty(payload.skillId, "skill id required");
  if (payload.prevManifestHash.length !== 32) throw new Error("invalid manifest hash length");
  decodeSkillManifest(payload.manifest);
  return {
    entries: [
      [0, payload.skillId],
      [1, payload.prevManifestHash],
      [2, payload.manifest],
      [3, payload.ts],
    ],
  } satisfies CborMap;
}

export function skillRevokePayloadToCbor(payload: SkillRevokePayload): CborValue {
  if (payload.ts === 0) throw new Error("timestamp required");
  ensureNonempty(payload.skillId, "skill id required");
  ensureNonempty(payload.reason, "reason required");
  if (payload.manifestHash.length !== 32) throw new Error("invalid manifest hash length");
  return {
    entries: [
      [0, payload.skillId],
      [1, payload.manifestHash],
      [2, payload.reason],
      [3, payload.ts],
    ],
  } satisfies CborMap;
}

function payloadToCbor(payload: SkillManifestPayload): CborValue {
  validatePayload(payload);
  const entries: [CborValue, CborValue][] = [
    [0, payload.skillId],
    [1, payload.author],
    [2, payload.name],
    [3, payload.version],
    [4, payload.summary],
    [5, payload.license],
    [6, [...payload.capabilities]],
    [7, [...payload.permissions]],
    [8, payload.sandboxClass],
    [15, payload.ts],
  ];
  if (payload.endpoints) {
    entries.push([9, [...payload.endpoints]]);
  }
  if (payload.artifacts) {
    entries.push([10, payload.artifacts.map(artifactToCbor)]);
  }
  if (payload.requirements) {
    entries.push([11, [...payload.requirements]]);
  }
  if (payload.pricing) {
    entries.push([12, payload.pricing]);
  }
  if (payload.attestations) {
    entries.push([13, payload.attestations]);
  }
  if (payload.metadata) {
    entries.push([14, payload.metadata]);
  }
  return { entries } satisfies CborMap;
}

function artifactToCbor(artifact: SkillArtifact): CborValue {
  validateArtifact(artifact);
  return { entries: [[0, artifact.kind], [1, artifact.digest], [2, artifact.size], [3, [...artifact.uris]]] } satisfies CborMap;
}

function parseArtifacts(value: CborValue): SkillArtifact[] {
  if (!Array.isArray(value) || value.length === 0) {
    throw new Error("artifacts required");
  }
  return value.map(parseArtifact);
}

function parseArtifact(value: CborValue): SkillArtifact {
  const entries = expectMap(value);
  const artifact: SkillArtifact = {
    kind: expectU8(getRequired(entries, 0)),
    digest: expectBytesLen(getRequired(entries, 1), 32),
    size: expectU64(getRequired(entries, 2)),
    uris: expectTextArray(getRequired(entries, 3)),
  };
  validateArtifact(artifact);
  return artifact;
}

function validatePayload(payload: SkillManifestPayload): void {
  ensureNonempty(payload.skillId, "skill id required");
  ensureNonempty(payload.author, "author required");
  ensureNonempty(payload.name, "name required");
  ensureNonempty(payload.version, "version required");
  ensureNonempty(payload.summary, "summary required");
  ensureNonempty(payload.license, "license required");
  ensureListNonempty(payload.capabilities, "capabilities required");
  ensureListItems(payload.permissions, "permissions required");
  if (payload.sandboxClass < SANDBOX_MIN || payload.sandboxClass > SANDBOX_MAX) {
    throw new Error("invalid sandbox class");
  }
  if (payload.endpoints) {
    ensureListNonempty(payload.endpoints, "endpoints required");
  }
  if (payload.artifacts) {
    if (payload.artifacts.length === 0) throw new Error("artifacts required");
    payload.artifacts.forEach(validateArtifact);
  }
  if (payload.requirements) {
    ensureListItems(payload.requirements, "requirements required");
  }
  if (!payload.endpoints && !payload.artifacts) {
    throw new Error("skill requires endpoints or artifacts");
  }
}

function validateArtifact(artifact: SkillArtifact): void {
  if (artifact.kind <= 0) throw new Error("artifact kind required");
  if (artifact.digest.length !== 32) throw new Error("artifact digest must be 32 bytes");
  if (artifact.size <= 0) throw new Error("artifact size required");
  ensureListNonempty(artifact.uris, "artifact uris required");
}

function ensureNonempty(value: string, message: string): void {
  if (!value || value.trim().length === 0) {
    throw new Error(message);
  }
}

function ensureListNonempty(values: string[], message: string): void {
  if (!values || values.length === 0) {
    throw new Error(message);
  }
  ensureListItems(values, message);
}

function ensureListItems(values: string[], message: string): void {
  for (const item of values) {
    if (!item || item.trim().length === 0) {
      throw new Error(message);
    }
  }
}

function splitSignedMap(value: CborValue, sigKey: number): [CborValue, Uint8Array] {
  const entries = expectMap(value);
  const payloadEntries: [CborValue, CborValue][] = [];
  let signature: Uint8Array | null = null;
  for (const [k, v] of entries) {
    const key = toNumber(k);
    if (typeof key === "number" && key === sigKey) {
      if (signature) throw new Error("duplicate signature key");
      if (!(v instanceof Uint8Array)) throw new Error("signature must be bytes");
      signature = v;
      continue;
    }
    payloadEntries.push([k, v]);
  }
  if (!signature) throw new Error("missing signature");
  if (signature.length !== 64) throw new Error("invalid signature length");
  return [{ entries: payloadEntries }, signature];
}

function withSignature(payload: CborValue, sigKey: number, signature: Uint8Array): CborValue {
  const entries = expectMap(payload);
  return { entries: [...entries, [sigKey, signature]] } satisfies CborMap;
}

function expectMap(value: CborValue): [CborValue, CborValue][] {
  if (typeof value === "object" && value !== null && "entries" in value) {
    return (value as CborMap).entries;
  }
  throw new Error("expected map");
}

function getRequired(entries: [CborValue, CborValue][], key: number): CborValue {
  for (const [k, v] of entries) {
    const idx = toNumber(k);
    if (typeof idx === "number" && idx === key) {
      return v;
    }
  }
  throw new Error("missing required key");
}

function getOptional(entries: [CborValue, CborValue][], key: number): CborValue | null {
  for (const [k, v] of entries) {
    const idx = toNumber(k);
    if (typeof idx === "number" && idx === key) {
      return v;
    }
  }
  return null;
}

function expectText(value: CborValue): string {
  if (typeof value === "string") return value;
  throw new Error("expected text");
}

function expectTextArray(value: CborValue): string[] {
  if (Array.isArray(value)) return value.map(expectText);
  throw new Error("expected array of text");
}

function expectBytesLen(value: CborValue, length: number): Uint8Array {
  if (!(value instanceof Uint8Array)) throw new Error("expected bytes");
  if (value.length !== length) throw new Error("invalid length");
  return value;
}

function expectBytes(value: CborValue): Uint8Array {
  if (!(value instanceof Uint8Array)) throw new Error("expected bytes");
  return value;
}

function expectU64(value: CborValue): number {
  const num = toNumber(value);
  if (typeof num !== "number" || num < 0 || !Number.isSafeInteger(num)) {
    throw new Error("expected unsigned");
  }
  return num;
}

function expectU16(value: CborValue): number {
  const num = toNumber(value);
  if (typeof num !== "number" || num < 0 || num > 0xffff) {
    throw new Error("expected u16");
  }
  return num;
}

function expectU8(value: CborValue): number {
  const num = toNumber(value);
  if (typeof num !== "number" || num < 0 || num > 0xff) {
    throw new Error("expected u8");
  }
  return num;
}

function toNumber(value: CborValue): number | null {
  if (typeof value === "number") return value;
  if (typeof value === "bigint") {
    const num = Number(value);
    if (!Number.isSafeInteger(num)) throw new Error("integer overflow");
    return num;
  }
  return null;
}

File: impl/python/agentnet_py/skill.py
from __future__ import annotations

from dataclasses import dataclass
from typing import List, Optional, Tuple

from .cbor import CborMap, CborValue, decode_canonical, encode_canonical
from .crypto import sha256, verify_ed25519_hash
from .sign import sign_ed25519_hash

SKILL_SIG_KEY = 16
SANDBOX_MIN = 1
SANDBOX_MAX = 5


class SkillError(ValueError):
    pass


@dataclass
class SkillArtifact:
    kind: int
    digest: bytes
    size: int
    uris: List[str]


@dataclass
class SkillManifestPayload:
    skill_id: str
    author: str
    name: str
    version: str
    summary: str
    license: str
    capabilities: List[str]
    permissions: List[str]
    sandbox_class: int
    endpoints: Optional[List[str]]
    artifacts: Optional[List[SkillArtifact]]
    requirements: Optional[List[str]]
    pricing: Optional[CborValue]
    attestations: Optional[CborValue]
    metadata: Optional[CborValue]
    ts: int


@dataclass
class SkillManifest:
    payload: SkillManifestPayload
    signature: bytes


@dataclass
class SkillPublishPayload:
    manifest: bytes
    ts: int


@dataclass
class SkillUpdatePayload:
    skill_id: str
    prev_manifest_hash: bytes
    manifest: bytes
    ts: int


@dataclass
class SkillRevokePayload:
    skill_id: str
    manifest_hash: bytes
    reason: str
    ts: int


def parse_skill_manifest_payload(value: CborValue) -> SkillManifestPayload:
    entries = _expect_map(value)
    payload = SkillManifestPayload(
        skill_id=_expect_text(_get_required(entries, 0)),
        author=_expect_text(_get_required(entries, 1)),
        name=_expect_text(_get_required(entries, 2)),
        version=_expect_text(_get_required(entries, 3)),
        summary=_expect_text(_get_required(entries, 4)),
        license=_expect_text(_get_required(entries, 5)),
        capabilities=_expect_text_array(_get_required(entries, 6)),
        permissions=_expect_text_array(_get_required(entries, 7)),
        sandbox_class=_expect_u16(_get_required(entries, 8)),
        endpoints=_optional_text_array(entries, 9),
        artifacts=_optional_artifacts(entries, 10),
        requirements=_optional_text_array(entries, 11),
        pricing=_get_optional(entries, 12),
        attestations=_get_optional(entries, 13),
        metadata=_get_optional(entries, 14),
        ts=_expect_u64(_get_required(entries, 15)),
    )
    _validate_payload(payload)
    return payload


def parse_skill_manifest(value: CborValue) -> SkillManifest:
    payload, signature = _split_signed_map(value, SKILL_SIG_KEY)
    payload_obj = parse_skill_manifest_payload(payload)
    return SkillManifest(payload=payload_obj, signature=signature)


def decode_skill_manifest(data: bytes) -> SkillManifest:
    return parse_skill_manifest(decode_canonical(data))


def build_skill_manifest(payload: SkillManifestPayload, secret_key: bytes) -> bytes:
    _validate_payload(payload)
    payload_cbor = _skill_payload_to_cbor(payload)
    payload_bytes = encode_canonical(payload_cbor)
    digest = sha256(payload_bytes)
    signature = sign_ed25519_hash(secret_key, digest)
    full = _with_signature(payload_cbor, SKILL_SIG_KEY, signature)
    return encode_canonical(full)


def verify_skill_manifest(data: bytes, public_key: bytes) -> SkillManifestPayload:
    value = decode_canonical(data)
    payload, signature = _split_signed_map(value, SKILL_SIG_KEY)
    payload_bytes = encode_canonical(payload)
    digest = sha256(payload_bytes)
    verify_ed25519_hash(public_key, digest, signature)
    return parse_skill_manifest_payload(payload)


def parse_skill_publish_payload(value: CborValue) -> SkillPublishPayload:
    entries = _expect_map(value)
    manifest = _expect_bytes(_get_required(entries, 0))
    ts = _expect_u64(_get_required(entries, 1))
    if ts == 0:
        raise SkillError("timestamp required")
    decode_skill_manifest(manifest)
    return SkillPublishPayload(manifest=manifest, ts=ts)


def parse_skill_update_payload(value: CborValue) -> SkillUpdatePayload:
    entries = _expect_map(value)
    skill_id = _expect_text(_get_required(entries, 0))
    prev_manifest_hash = _expect_bytes_len(_get_required(entries, 1), 32)
    manifest = _expect_bytes(_get_required(entries, 2))
    ts = _expect_u64(_get_required(entries, 3))
    if ts == 0:
        raise SkillError("timestamp required")
    decode_skill_manifest(manifest)
    return SkillUpdatePayload(
        skill_id=skill_id,
        prev_manifest_hash=prev_manifest_hash,
        manifest=manifest,
        ts=ts,
    )


def parse_skill_revoke_payload(value: CborValue) -> SkillRevokePayload:
    entries = _expect_map(value)
    skill_id = _expect_text(_get_required(entries, 0))
    manifest_hash = _expect_bytes_len(_get_required(entries, 1), 32)
    reason = _expect_text(_get_required(entries, 2))
    ts = _expect_u64(_get_required(entries, 3))
    if ts == 0:
        raise SkillError("timestamp required")
    if reason.strip() == "":
        raise SkillError("reason required")
    return SkillRevokePayload(
        skill_id=skill_id,
        manifest_hash=manifest_hash,
        reason=reason,
        ts=ts,
    )


def skill_publish_payload_to_cbor(payload: SkillPublishPayload) -> CborValue:
    if payload.ts == 0:
        raise SkillError("timestamp required")
    decode_skill_manifest(payload.manifest)
    return CborMap([(0, bytes(payload.manifest)), (1, payload.ts)])


def skill_update_payload_to_cbor(payload: SkillUpdatePayload) -> CborValue:
    if payload.ts == 0:
        raise SkillError("timestamp required")
    _ensure_nonempty(payload.skill_id, "skill id required")
    if len(payload.prev_manifest_hash) != 32:
        raise SkillError("invalid manifest hash length")
    decode_skill_manifest(payload.manifest)
    return CborMap(
        [
            (0, payload.skill_id),
            (1, bytes(payload.prev_manifest_hash)),
            (2, bytes(payload.manifest)),
            (3, payload.ts),
        ]
    )


def skill_revoke_payload_to_cbor(payload: SkillRevokePayload) -> CborValue:
    if payload.ts == 0:
        raise SkillError("timestamp required")
    _ensure_nonempty(payload.skill_id, "skill id required")
    _ensure_nonempty(payload.reason, "reason required")
    if len(payload.manifest_hash) != 32:
        raise SkillError("invalid manifest hash length")
    return CborMap(
        [
            (0, payload.skill_id),
            (1, bytes(payload.manifest_hash)),
            (2, payload.reason),
            (3, payload.ts),
        ]
    )


def _skill_payload_to_cbor(payload: SkillManifestPayload) -> CborValue:
    _validate_payload(payload)
    entries: List[Tuple[CborValue, CborValue]] = [
        (0, payload.skill_id),
        (1, payload.author),
        (2, payload.name),
        (3, payload.version),
        (4, payload.summary),
        (5, payload.license),
        (6, list(payload.capabilities)),
        (7, list(payload.permissions)),
        (8, payload.sandbox_class),
        (15, payload.ts),
    ]
    if payload.endpoints is not None:
        entries.append((9, list(payload.endpoints)))
    if payload.artifacts is not None:
        entries.append((10, [_artifact_to_cbor(a) for a in payload.artifacts]))
    if payload.requirements is not None:
        entries.append((11, list(payload.requirements)))
    if payload.pricing is not None:
        entries.append((12, payload.pricing))
    if payload.attestations is not None:
        entries.append((13, payload.attestations))
    if payload.metadata is not None:
        entries.append((14, payload.metadata))
    return CborMap(entries)


def _artifact_to_cbor(artifact: SkillArtifact) -> CborValue:
    _validate_artifact(artifact)
    return CborMap(
        [
            (0, artifact.kind),
            (1, bytes(artifact.digest)),
            (2, artifact.size),
            (3, list(artifact.uris)),
        ]
    )


def _optional_text_array(entries: List[Tuple[CborValue, CborValue]], key: int) -> Optional[List[str]]:
    value = _get_optional(entries, key)
    if value is None:
        return None
    return _expect_text_array(value)


def _optional_artifacts(entries: List[Tuple[CborValue, CborValue]], key: int) -> Optional[List[SkillArtifact]]:
    value = _get_optional(entries, key)
    if value is None:
        return None
    if not isinstance(value, list):
        raise SkillError("expected artifact array")
    if len(value) == 0:
        raise SkillError("artifacts required")
    artifacts: List[SkillArtifact] = []
    for item in value:
        artifacts.append(_parse_artifact(item))
    return artifacts


def _parse_artifact(value: CborValue) -> SkillArtifact:
    entries = _expect_map(value)
    artifact = SkillArtifact(
        kind=_expect_u8(_get_required(entries, 0)),
        digest=_expect_bytes_len(_get_required(entries, 1), 32),
        size=_expect_u64(_get_required(entries, 2)),
        uris=_expect_text_array(_get_required(entries, 3)),
    )
    _validate_artifact(artifact)
    return artifact


def _validate_payload(payload: SkillManifestPayload) -> None:
    _ensure_nonempty(payload.skill_id, "skill id required")
    _ensure_nonempty(payload.author, "author required")
    _ensure_nonempty(payload.name, "name required")
    _ensure_nonempty(payload.version, "version required")
    _ensure_nonempty(payload.summary, "summary required")
    _ensure_nonempty(payload.license, "license required")
    _ensure_list_nonempty(payload.capabilities, "capabilities required")
    _ensure_list_items(payload.permissions, "permissions required")
    if payload.sandbox_class < SANDBOX_MIN or payload.sandbox_class > SANDBOX_MAX:
        raise SkillError("invalid sandbox class")
    if payload.endpoints is not None:
        _ensure_list_nonempty(payload.endpoints, "endpoints required")
    if payload.artifacts is not None:
        if len(payload.artifacts) == 0:
            raise SkillError("artifacts required")
        for artifact in payload.artifacts:
            _validate_artifact(artifact)
    if payload.requirements is not None:
        _ensure_list_items(payload.requirements, "requirements required")
    if payload.endpoints is None and payload.artifacts is None:
        raise SkillError("skill requires endpoints or artifacts")


def _validate_artifact(artifact: SkillArtifact) -> None:
    if artifact.kind <= 0:
        raise SkillError("artifact kind required")
    if len(artifact.digest) != 32:
        raise SkillError("artifact digest must be 32 bytes")
    if artifact.size <= 0:
        raise SkillError("artifact size required")
    _ensure_list_nonempty(artifact.uris, "artifact uris required")


def _ensure_nonempty(value: str, message: str) -> None:
    if value is None or value.strip() == "":
        raise SkillError(message)


def _ensure_list_nonempty(values: List[str], message: str) -> None:
    if len(values) == 0:
        raise SkillError(message)
    _ensure_list_items(values, message)


def _ensure_list_items(values: List[str], message: str) -> None:
    for item in values:
        if item is None or str(item).strip() == "":
            raise SkillError(message)


def _split_signed_map(value: CborValue, sig_key: int) -> Tuple[CborValue, bytes]:
    entries = _expect_map(value)
    payload_entries: List[Tuple[CborValue, CborValue]] = []
    signature: Optional[bytes] = None
    for key, val in entries:
        if key == sig_key:
            if signature is not None:
                raise SkillError("duplicate signature key")
            if not isinstance(val, (bytes, bytearray)):
                raise SkillError("signature must be bytes")
            signature = bytes(val)
            continue
        payload_entries.append((key, val))
    if signature is None:
        raise SkillError("missing signature")
    if len(signature) != 64:
        raise SkillError("invalid signature length")
    return CborMap(payload_entries), signature


def _with_signature(payload: CborValue, sig_key: int, signature: bytes) -> CborValue:
    entries = list(_expect_map(payload))
    entries.append((sig_key, signature))
    return CborMap(entries)


def _expect_map(value: CborValue) -> List[Tuple[CborValue, CborValue]]:
    if isinstance(value, CborMap):
        return value.entries
    raise SkillError("expected map")


def _get_required(entries: List[Tuple[CborValue, CborValue]], key: int) -> CborValue:
    for k, v in entries:
        if k == key:
            return v
    raise SkillError("missing required key")


def _get_optional(entries: List[Tuple[CborValue, CborValue]], key: int) -> Optional[CborValue]:
    for k, v in entries:
        if k == key:
            return v
    return None


def _expect_text(value: CborValue) -> str:
    if isinstance(value, str):
        return value
    raise SkillError("expected text")


def _expect_text_array(value: CborValue) -> List[str]:
    if isinstance(value, list):
        out: List[str] = []
        for item in value:
            out.append(_expect_text(item))
        return out
    raise SkillError("expected array of text")


def _expect_bytes_len(value: CborValue, length: int) -> bytes:
    if isinstance(value, (bytes, bytearray)):
        data = bytes(value)
        if len(data) != length:
            raise SkillError("invalid length")
        return data
    raise SkillError("expected bytes")


def _expect_bytes(value: CborValue) -> bytes:
    if isinstance(value, (bytes, bytearray)):
        return bytes(value)
    raise SkillError("expected bytes")


def _expect_u64(value: CborValue) -> int:
    if isinstance(value, int) and value >= 0:
        return value
    raise SkillError("expected unsigned")


def _expect_u16(value: CborValue) -> int:
    if isinstance(value, int) and 0 <= value <= 0xFFFF:
        return value
    raise SkillError("expected u16")


def _expect_u8(value: CborValue) -> int:
    if isinstance(value, int) and 0 <= value <= 0xFF:
        return value
    raise SkillError("expected u8")
